### PR TITLE
Users/ealsur/v3bulk

### PR DESCRIPTION
**Pull request by TheovanKraay**
Change to add the ability to configure mapping of an attribute in source collection (or array of attributes for a synthetic key) to a field designated as the partition key for the target collection.

### CODE REVIEW

File: src/DocumentFeedMigrator.cs [hunk 1]
```diff
@@ -4,22 +4,26 @@
 
 namespace MigrationExecutorFunctionApp
 {
+    using Microsoft.Azure.Cosmos;
```
Comment (ealsur) [inline src/DocumentFeedMigrator.cs]: Order. System always comes first.

File: src/DocumentFeedMigrator.cs [hunk 2]
```diff
@@ -67,10 +70,24 @@ public DocumentFeedMigrator(Container containerToStoreDocuments)
...
+
+        public Document MapPartitionKey(Document document, Document doc)
+        {
+            dynamic json = JsonConvert.DeserializeObject(doc.ToString());
+            StringBuilder syntheticKey = new StringBuilder();
+            string[] sourceAttributeArray = sourcePartitionKeyMapping.Split(',');
+            foreach (string attribute in sourceAttributeArray) {
+                syntheticKey.Append(json[attribute]);
+            }
+            json[targetPartitionKeyAttribute] = syntheticKey.ToString();
```
Comment (ealsur) [inline src/DocumentFeedMigrator.cs]: What about just using document.GetPropertyValue and document.SetPropertyValue, that removes all the dynamic JSON parsing.

File: src/local.settings.json [hunk 3]
```diff
@@ -4,13 +4,15 @@
...
+    "SourceCosmosDB": "",
+    "SourceDatabase": "database",
+    "SourceCollection": "container1",
+    "TargetCosmosDB": "",
+    "TargetDatabase": "database",
+    "TargetCollection": "container2",
+    "TargetPartitionKeyAttribute": "partitionKey",
+    "SourcePartitionKeyMapping": "deviceid,timestamp",
+    "QueueName": "queue",
+    "QueueConnectionString": "DefaultEndpointsProtocol=https;AccountName=<storage account name>;AccountKey=<storage account key>;QueueEndpoint=https://<storage account name>.queue.core.windows.net"
```
Comment (ealsur) [inline src/local.settings.json]: Basic storage connection does not require queue endpoint, just account name and key

File: src/DocumentFeedMigrator.cs [hunk 4]
```diff
@@ -4,22 +4,27 @@
...
+
+
 
     public class DocumentFeedMigrator
     {
         private Container containerToStoreDocuments;
-
+        string targetPartitionKeyAttribute = Environment.GetEnvironmentVariable($"{"TargetPartitionKeyAttribute"}");
+        static string sourcePartitionKeyMapping = Environment.GetEnvironmentVariable($"{"SourcePartitionKeyMapping"}");    
+        Boolean isSyntheticKey = Environment.GetEnvironmentVariable($"{"SourcePartitionKeyMapping"}").IndexOf(",") == -1? false: true;
```
Comment (ealsur) [inline src/DocumentFeedMigrator.cs:27]: With this change you are now extracting the PK even on a scenario where the destination PK is not synthetic, which doesn't add any value.

Comment (ealsur) [inline src/DocumentFeedMigrator.cs:27]: Shouldn't be checking the TargetPartitionKeyAttribute?

File: src/DocumentFeedMigrator.cs [hunk 5]
```diff
@@ -67,10 +71,45 @@ public DocumentFeedMigrator(Container containerToStoreDocuments)
...
+            {
+                doc.SetPropertyValue(targetPartitionKeyAttribute, doc.GetPropertyValue<string>(sourcePartitionKeyMapping));
+            }
+            return doc;
+        }
+
+        public Document CreateSyntheticKey(Document doc)
+        {
+            StringBuilder syntheticKey = new StringBuilder();
+            string[] sourceAttributeArray = sourcePartitionKeyMapping.Split(',');
```
Comment (ealsur) [inline src/DocumentFeedMigrator.cs:96]: I wonder, will this work with nested PKs? like `property.nestedProperty`?

File: src/local.settings.json [hunk 6]
```diff
@@ -4,13 +4,15 @@
...
     "FUNCTIONS_WORKER_RUNTIME": "dotnet",
-    "SourceCosmosDB": "<Your connection string to your Monitored Container database account here>",
-    "SourceDatabase": "<your Monitored Container's database name here>",
-    "SourceCollection": "<your Monitored Container's name here>",
-    "TargetCosmosDB": "<Your connection string to your Target Container databse account here>",
-    "TargetDatabase": "<your Target Container's database name here>",
-    "TargetCollection": "<your Target Container's name here>",
-    "PoisonQueueName": "<name of the Queue to store poison documents>",
-    "PoisonQueueConnectionString": "<Azure Storage Connection String to hold the poison documents Queue>"
+    "SourceCosmosDB": "",
```
Comment (ealsur) [inline src/local.settings.json]: We might want to leave the previous value here or a comment, not an empty value?

File: src/DocumentFeedMigrator.cs (hunk 4 above)
Comment (TheovanKraay) [reply src/DocumentFeedMigrator.cs:27]: I only need to check whether there are many attributes, or one, in the source partition key field. For TargetPartitionKeyAttribute I am assuming they provide a valid single field. Perhaps we could add a check. But these fields need documenting in the readme, which I have not done yet.

File: src/local.settings.json (hunk 6 above)
Comment (TheovanKraay) [reply src/local.settings.json]: yes, I will add.

File: src/DocumentFeedMigrator.cs (hunk 5 above)
Comment (TheovanKraay) [reply src/DocumentFeedMigrator.cs:96]: I am not sure, but in practice I think nested pk is rare.

File: src/local.settings.json (hunk 3 above)
Comment (ealsur) [reply src/local.settings.json]:
```suggestion
    "QueueConnectionString": "DefaultEndpointsProtocol=https;AccountName=<storage account name>;AccountKey=<storage account key>;"
```

**Pull request merged**